sandbox/cgroup/monitor.go

```diff
@@ -33,122 +33,190 @@ import (
-	wd:           nil,
-	pathList:     make(map[string]int32),
-	addWatchChan: make(chan *groupToWatch),
+var currentWatcher *inotifyWatcher = newInotifyWatcher(context.Background())
+
+func newInotifyWatcher(ctx context.Context) *inotifyWatcher {
+	ctx, cancel := context.WithCancel(ctx)
+	return &inotifyWatcher{
+		ctx:          ctx,
+		pathList:     make(map[string]uint),
+		addWatchChan: make(chan *groupToWatch),
+		doneChan:     make(chan struct{}),
+		done:         cancel,
+	}
 }
 
 func (gtw *groupToWatch) sendClosedNotification() {
-	defer func() {
-		if err := recover(); err != nil {
-			logger.Noticef("Failed to send Closed notification for %s", gtw.name)
-		}
+	go func() {
+		defer func() {
+			if err := recover(); err != nil {
+				logger.Noticef("cannot send closed notification for %s", gtw.name)
+			}
+		}()
+		gtw.channel <- gtw.name
 	}()
-	gtw.channel <- gtw.name
+}
+
+// Close stops the watcher and waits for it to finish.
+func (iw *inotifyWatcher) Close() {
+	iw.done()
+	<-iw.doneChan
+	for p := range iw.pathList {
+		iw.removePath(p)
+	}
+	iw.wd.Close()
 }
 
 func (iw *inotifyWatcher) addWatch(newWatch *groupToWatch) {
-	var folderList []string
-	for _, fullPath := range newWatch.folders {
+	for fullPath := range newWatch.folders {
 		// It's not possible to use inotify.InDeleteSelf in /sys/fs because it
 		// isn't triggered, so we must monitor the parent folder and use InDelete
-		basePath := path.Dir(fullPath)
+		basePath := filepath.Dir(fullPath)
 		if _, exists := iw.pathList[basePath]; !exists {
 			iw.pathList[basePath] = 0
 			if err := iw.wd.AddWatch(basePath, inotify.InDelete); err != nil {
-				target := &os.PathError{}
-				if !errors.As(err, &target) {
-					logger.Noticef("Error when calling AddWatch for path %s: %s", target.Path, target.Err)
-				}
+				// TODO propagate the error back to the caller
+				logger.Noticef("cannot add watch %v", err)
 				delete(iw.pathList, basePath)
 				continue
 			}
 		}
+
+		// bump for the base path, since we're relying on a watch being added there
+		iw.pathList[basePath]++
+		// bump on the actual path
 		iw.pathList[fullPath]++
-		if osutil.FileExists(fullPath) {
-			folderList = append(folderList, fullPath)
-		} else {
+
+		if !osutil.FileExists(fullPath) {
+			// the path is gone by now
+			delete(newWatch.folders, fullPath)
 			iw.removePath(fullPath)
 		}
 	}
-	if len(folderList) == 0 {
+
+	if len(newWatch.folders) == 0 {
 		newWatch.sendClosedNotification()
 	} else {
-		newWatch.folders = folderList
 		iw.groupList = append(iw.groupList, newWatch)
 	}
+
+	iw.notifyObserver(newWatch)
+
+	logger.Debugf("watches after add: %v", iw.pathList)
 }
 
 func (iw *inotifyWatcher) removePath(fullPath string) {
-	iw.pathList[fullPath]--
-	if iw.pathList[fullPath] == 0 {
-		iw.wd.RemoveWatch(fullPath)
-		delete(iw.pathList, fullPath)
+	cnt, ok := iw.pathList[fullPath]
+
+	if !ok {
+		// path we are not watching
+		return
+	}
+
+	parent := filepath.Dir(fullPath)
```

Comment:
I think this logic will misbehave is someone uses it to monitor `/`.
Perhaps add some guard to avoid doing that and avoid making a negative `pathList` counter below.